src/viability_phase.py

```diff
@@ -0,0 +1,163 @@
+from api import ml
+from api import al
+from sklearn.metrics import mean_squared_error
+from tqdm import tqdm
+import numpy as np
+
+
+def viability_phase(learning_profiles: list, num_iterations: int = -1):
+    """
+    Performs active learning for all Learning Profiles. However, instead of
+    querying the user, pre-labeled data will be used for selected samples.
+
+    Args:
+        learning_profiles (list): A list of all Learning Profiles to train.
+        num_iterations (int): Number of evaluation iterations. Set to -1 if
+            all training songs should be depleted. Default is -1.
+    """
+
+    # Perform viability test for all learning profiles
+    for lp in learning_profiles:
+        print(f"Evaluating {lp}")
+        MSE_arousal, MSE_valence = _evaluate(lp, num_iterations)
+        lp.set_MSE(MSE_arousal, MSE_valence)
+
+        print(f"Done!")
+        print(f"Arousal MSE per iteration: {lp.get_MSE_arousal()}")
+        print(f"Valence MSE per iteration: {lp.get_MSE_valence()}")
+
+
+def _evaluate(lp: LearningProfile, num_iterations: int):
+    """
+    Performs the following steps for the given learning profile:
+        1: Add first song from training dataset to a seed dataset.
+        2: Perform ML on this song to predict arousal/valence.
+        3: Perform AL to find novel songs in training dataset to
+            add to the seed dataset.
+        4: Add these songs to the seed dataset.
+        5: Fit ML model to the seed dataset and predict on test dataset
+            to get MSE values for current iteration.
+        6: Repeat 3-5 for a specified number of iterations: `num_iterations`.
+
+    Args:
+        lp (LearningProfile): Learning profile to evaluate.
+        num_iterations (int): Number of evaluation iterations. Set to -1 if
+            all training songs should be depleted.
+
+    Returns:
+        (np.ndarray, np.ndarray): Arrays with arousal and valence MSE values.
+    """
+
+    ########
+    # Init #
+    ########
+
+    # Get functions
+    al_func = lp.get_al_function()
+    ml_func = lp.get_ml_function()
+
+    # Get full datasets
+    full_train_ds = lp.get_train_dataset()
+    full_test_ds = lp.get_test_dataset()
+
+    # Get song ids of full training data
+    full_train_song_ids = full_train_ds.get_contained_song_ids()
+
+    ##########
+    # Step 1 #
+    ##########
+
+    # Initialize training songs with first song id
+    seed_song_ids = [full_train_ds.get_data().iloc[0].index.values[0]]
+
+    # Mark all but the first song as unlabeled
+    unlabeled_train_song_ids = list(
+        set(full_train_song_ids).difference(seed_song_ids))
+
+    # Get dataframe and labels of full training data
+    full_train_dat = full_train_ds.get_data()
+    full_train_ar = full_train_ds.get_arousal_labels()
+    full_train_va = full_train_ds.get_valence_labels()
+
+    # Derive number of samples per song (assume same for all songs)
+    n_samples_per_song = full_train_dat.loc[seed_song_ids, :].shape[0]
```

Comment:
seed_song_ids[0] -> to make sure one song is selected

Reply:
Not necessary.